actor/storage.go

```diff
@@ -38,27 +38,49 @@ func UnmarshalStorage(raw []byte, to interface{}) error {
 // Note that if 'f' returns an error, modifications to the storage are not
 // saved.
 func WithState(ctx exec.VMContext, st interface{}, f func() (interface{}, error)) (interface{}, error) {
-	chunk, err := ctx.ReadStorage()
-	if err != nil {
-		return nil, vmerrors.FaultErrorWrap(err, "Could not read actor storage")
-	}
-
-	if err := UnmarshalStorage(chunk, st); err != nil {
-		return nil, vmerrors.FaultErrorWrap(err, "Could not unmarshall actor storage")
+	if err := ReadState(ctx, st); err != nil {
+		return nil, err
 	}
 
 	ret, err := f()
 	if err != nil {
 		return nil, err
 	}
 
-	if err := ctx.WriteStorage(st); err != nil {
-		return nil, vmerrors.FaultErrorWrap(err, "Could not write actor storage")
+	stage := ctx.Storage()
+
+	cid, err := stage.Put(st)
+	if err != nil {
+		return nil, vmerrors.RevertErrorWrap(err, "Could not stage memory chunk")
+	}
+
+	err = stage.Commit(cid, stage.Head())
+	if err != nil {
+		return nil, vmerrors.RevertErrorWrap(err, "Could not commit actor memory")
 	}
 
 	return ret, nil
 }
 
+// ReadState is a helper method to read the cbor node at the actor's Head into the given struct
+func ReadState(ctx exec.VMContext, st interface{}) error {
+	storage := ctx.Storage()
+
+	memory, err := storage.Get(storage.Head())
+	if err != nil {
+		return vmerrors.FaultErrorWrap(err, "Could not read actor storage")
+	}
+
+	chunk := make([]byte, len(memory))
+	copy(chunk, memory)
```

Comment:
Could you explain why we need to copy here?  Looking at the `storage.Get` method's implementation it is not clear what the benefit is.

Reply:
I agree we don't. This is copied from really old code. I removed it.